### PR TITLE
refactor: use shard uid in flat state keys

### PR DESCRIPTION
**Pull request by Longarithm**
Prepend `ShardUId` to flat state keys, as motivated in https://github.com/near/nearcore/issues/8670.

Ideally we need to replace `shard_id` with `shard_uid` everywhere, including other DB keys and `RuntimeAdapter` flat storage methods. But this change would be too involved, so let's do it incrementally.

The biggest structural consequence is that `FlatStorage` now stores `shard_uid`, which makes sense, and later it will be used for writing deltas as well. It's not a big deal, but because of that changes are scattered throughout the code and may look scary.

## Testing

* `test_flat_storage_iter` checks that new `key_belongs_to_shard` impl is aligned with shard layout; also we can see that the change takes place in this failing test: https://buildkite.com/nearprotocol/nearcore/builds/25676#0186bc8f-12a2-410a-b74c-e326f3b963fa
* https://nayduck.near.org/#/run/2898

### CODE REVIEW

File: core/store/src/flat/storage.rs [hunk 1]
```diff
@@ -41,7 +41,7 @@ pub struct FlatStorage(pub(crate) Arc<RwLock<FlatStorageInner>>);
 pub(crate) struct FlatStorageInner {
     store: Store,
     /// Id of the shard which state is accessed by this flat storage.
```
Comment (pugachAG) [inline core/store/src/flat/storage.rs:43]: nit: update the comment above as well

File: core/store/src/flat/store_helper.rs [hunk 2]
```diff
@@ -97,9 +96,32 @@ pub fn remove_flat_head(store_update: &mut StoreUpdate, shard_id: ShardId) {
...
+}
+
+fn decode_flat_state_db_key(key: &Box<[u8]>) -> Result<(ShardUId, Vec<u8>), StorageError> {
+    if key.len() < 8 {
+        return Err(StorageError::StorageInconsistentState(format!(
+            "Found key in flat storage with length < 8: {key:?}"
+        )));
+    }
+    let shard_uid = key[0..8].try_into().unwrap();
+    let trie_key = key[8..].to_vec();
```
Comment (pugachAG) [inline core/store/src/flat/store_helper.rs:113]: nit: use [`split_at`](https://doc.rust-lang.org/std/primitive.slice.html#method.split_at) and avoid `unwrap`:
```
let (shard_uid_bytes, trie_key) = key.split_at(8);
let shard_uid = shard_uid_bytes.try_into().map_err(|_| StorageError::StorageInconsistentState(..))?;
Ok((shard_uid, trie_key.to_vec())
```

File: core/store/src/flat/store_helper.rs [hunk 3]
```diff
@@ -197,17 +222,18 @@ pub fn remove_flat_storage_creation_status(store_update: &mut StoreUpdate, shard
 }
 
 /// Iterate over flat storage entries for a given shard.
-/// It reads data only from the 'main' column - which represents the state as of final head.k
+/// It reads data only from the 'main' column - which represents the state as of final head.
 ///
 /// WARNING: flat storage keeps changing, so the results might be inconsistent, unless you're running
 /// this method on the shapshot of the data.
+// TODO(#8676): Support non-trivial ranges.
 pub fn iter_flat_state_entries<'a>(
```
Comment (pugachAG) [inline core/store/src/flat/store_helper.rs:230]: it looks like we also need to chage this method to accept `shard_uid` instead of `shard_id` and do a proper prefix-read? or do you please to do that as a separate PR?

**Comment by pugachAG** [comment]
also update `columns.rs`:
```
- DBCol::FlatState => &[DBKeyType::TrieKey],
+ DBCol::FlatState => &[DBKeyType::ShardUId, DBKeyType::TrieKey],
```

**Comment by akhi3030** [comment]
@pugachAG: please ping me when you are satisfied with this PR and it is ready to be merged.

File: core/store/src/flat/store_helper.rs (hunk 2 above)
Comment (Longarithm) [reply core/store/src/flat/store_helper.rs]: Nice!

File: core/store/src/flat/store_helper.rs (hunk 3 above)
Comment (Longarithm) [reply core/store/src/flat/store_helper.rs:234]: Yeah, I would do it separately. Maybe even outside MVP, because we don't use this logic much.

**Review by pugachAG — APPROVED** [review]
LGTM! @akhi3030

**Review by akhi3030 — APPROVED** [review]
